### ISSUE TITLE
Monaco API Request: Category Headers in Suggestions

### ISSUE THREAD

**Issue by tracksaw** [microsoft/vscode#67474]
In our DSL we know exactly what list of suggestions we want to show given the current user input and the domain profile of the user. 

The universe of possible suggestions is extremely large in this domain. The suggestions for given user input break down naturally into a few categories, and given the context we know which categories are appropriate. We also know, with quite high probability, the most likely 4 or 5 suggestions in each category. We therefore present the search results as a single list with headers representing the result categories. 

We prefer the headered list approach to the “item plus icon” display that is standard in VSCode because this is consistent with our other products that customers use on a daily basis.

Below is an example in the our DSL Editor, where the user has typed ‘m’.

### Feature Requested

 Allow Suggestion list items to be configured as headers, such that:
•	Header items in the list are identifiable for styling purposes.
•	The up and down arrow keys “hop over” header items.
•	The mouse click action for a header item is settable to do: nothing, a custom action (callback), or the default action.

![bqleditor-search-for-m](https://user-images.githubusercontent.com/29378846/51935940-19d66d00-23d5-11e9-85c7-31100d3af2dc.png)

**Comment by jrieken** [comment]
Leaving this open as discussion item but you should know that this is really not how our IntelliSense system is designed to work and that chances are low that we do this.

**Comment by jrieken** [comment]
TBH - if you using the monaco editor you are better off building your own IntelliSense UX and model as this doesn't fit with anything we have or plan.

**Comment by tracksaw** [comment]
We find Monaco profoundly useful for the kind of product we are building. One of the many things we like is the IntelliSense system. 

The suggestion we are making here is more to do with the presentation layer of IntelliSense (i.e., how it looks), than the functional layer (i.e., how it's meant to work).

For our purposes we want to group suggestions by category. When you do that with the  “item plus icon” display that is standard in VSCode the display looks repetitive and cluttered. We prefer the headered approach on that basis (and also for standardization with our other products).

BTW. We have implemented this display on a fork of the Monaco code base. Initial feedback from users is very positive. We think this display style could become quite popular, and we'd like to get it into the code base to contribute to the community and avoid maintaining a separate fork.

**Comment by jrieken** [comment]
> The suggestion we are making here is more to do with the presentation layer of IntelliSense (i.e., how it looks), than the functional layer (i.e., how it's meant to work).

You are also suggesting changes to how the model works, e.g. there are no categories today, there exists no concept of "More Something...". Today we have kinds, which are fixed, but items don't group by them but sort by match quality - independent of kind. Not sure how that works in your world. 

> We have implemented this display on a fork of the Monaco code base. Initial feedback from users is very positive.

That's very good for you but I don't know if that's universal feedback. The following sample

```ts
class Foo {
 fieldu: number;
 shieldu:number;
 fun() {}
 sun() {}
}

let a:Foo;

a.f // <IntelliSense>
```

Given the sample above a provider returns *all* members, then VS Code filters those that don't match and sorts the rest by match quality. In your model, what will happen? 

```
Properties
  fieldu //selected?
  sieldu
Methods
  fun
  sun
```
⬆️ this shows all and only selects the best 

```
Properties
  fieldu //selected?
Methods
  fun
```
⬆️ this filters but still cannot resort bested on match quality

What happens with `a.fu`, change the selection order reorder. What happens when I have another property `fuo`. Resort and regroup? Ignore the second best match?

```
Properties
  fuo //selected?
Methods
  fun //selected?
Properties
  fieldu // matches but weak
```
⬆️ repeat groups

```
Properties
  fuo //selected?
  fieldu // matches but weak
Methods
  fun //selected?
```
⬆️ weak matches on top of good matches?

So, I think this isn't a good fit for the "type to complete" use case as there is much visual noise and it increases the distance from the cursor and the completion item I want.

**Comment by jrieken** [comment]
related (duplicate) issues: https://github.com/Microsoft/vscode/issues/50755, https://github.com/Microsoft/vscode/issues/50291

**Comment by kieferrm** [comment]
@jrieken labeling this as feature-request to get this out of the inbox.

**Comment by tracksaw** [comment]
Thanks! This is a feature request.

> That's very good for you but I don't know if that's universal feedback. 

I said "initial feedback", not "universal feedback". I don't see how anyone could get "universal feedback" for a newly proposed feature.

> Given the sample above a provider returns all members, then VS Code filters those that don't match and sorts the rest by match quality. In your model, what will happen?

We have a single suggest provider in use and we turned off sorting and filtering by disabling it in code in response to a configuration option. Because of that, the items and groups are displayed, unfiltered, in the exact order we provide them. That is the intent of this feature.

How that would interact with another provider in use simultaneously, I don't know. I can think of plenty of ways of resolving that issue.

> ... I think this isn't a good fit for the "type to complete" use case as there is much visual noise and it increases the distance from the cursor and the completion item I want.

This is a DSL that operates in a domain in which we are fairly expert. While superficially the user of the DSL editor is editing "code", the language has far more keywords than a typical programming language but they cluster into a fairly small number of groups. We know from context what groups are likely. Also the text the user types may be quite different from the appropriate suggestion. (Consider typing "mammal" to get a list like: dog, cat, gerbil.) We know these relationships and can make sensible suggestions under these circumstances. 

In short our matching algorithm is about as good as you'll find, which is why we want our results displayed verbatim.

The contrib/suggest framework provides just about everything we need to display these results, except the ability to disable sorting and filtering. I guess we could make our own contrib for this. It would basically be a clone of suggest with a boolean config flag and the tiny code change to disable sorting and filtering.

Would that be your recommended approach?

Assuming it was sensibly implemented would you accept something like that as a contrib?

**Comment by jrieken** [comment]
> Assuming it was sensibly implemented would you accept something like that as a contrib?

Unlikely tbh - we would have the cost of maintaining it and this isn't upvoted or matching the existing vs code suggestion logic and API.
